From b7f85e4b35aeead9b3c64f06c0c4a4617baaf7f5 Mon Sep 17 00:00:00 2001
From: Anssi Hannula <anssi.hannula@iki.fi>
Date: Sat, 9 Feb 2013 01:37:48 +0200
Subject: [PATCH] hcontrol: Add workaround for handling IEC958 workaround

IEC958 workaround in the kernel can cause controls that have same the
same name and index but a different device.

Temporarily ignore failure of adding such conflicting controls if the
mixer class fails on them.
---
 src/control/hcontrol.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/control/hcontrol.c b/src/control/hcontrol.c
index ee1d907..dfc0a5a 100644
--- a/src/control/hcontrol.c
+++ b/src/control/hcontrol.c
@@ -612,8 +612,13 @@ int snd_hctl_load(snd_hctl_t *hctl)
 	for (idx = 0; idx < hctl->count; idx++) {
 		int res = snd_hctl_throw_event(hctl, SNDRV_CTL_EVENT_MASK_ADD,
 					       hctl->pelems[idx]);
-		if (res < 0)
-			return res;
+		if (res < 0) {
+			if (res == -EINVAL && snd_hctl_elem_get_device(hctl->pelems[idx]) == 1) {
+				SNDERR("ignoring mixer element addition failure on device 1 - IEC958 conflict workaround");
+			} else {
+				return res;
+			}
+		}
 	}
 	err = snd_ctl_subscribe_events(hctl->ctl, 1);
  _end:
-- 
1.7.10

